### DISCUSSION TITLE
does not build out of the box on MacOS

### DISCUSSION THREAD

**Discussion by jlangr** [JetBrains/ideavim#470] [General]
Greetings!

I'd like to start helping add plugin emulations to the ideavim plugin.

After forking and cloning to my machine, if I try to run the tests (using the IDEA run configuration `IdeaVim tests`), I end up with 
a failure report that indicates 260 failures out of 457 tests. Most if not all of them appear to be:

`java.lang.NoClassDefFoundError: Could not initialize class org.jetbrains.plugins.ideavim.VimTestCase`

Running the full verification IDEA run config shows 2 errors, and it looks like it's indicating a bunch of lint failures. A command-line `./gradlew build` shows the same two sets of lint errors, but does not appear to fail otherwise.

Running `./gradlew test` shows:
  `2866 tests completed, 2669 failed`

Most if not all of these appear to be:
` java.nio.file.FileSystemException:   /private/var/folders/ls/vd4q6fms761_nwwpz43n_8xm0000gn/T/unitTest__simple_action_25bxvoGfbZUYZz5xk5Z0QxP1Mk8/unitTest__unregister [[... lots more here... ]]: File name too long`

I'm running on MacOS (Big Sur) and in IDEA 2021.2.3 (Ultimate Edition), with open JDK 17 specified as the JDK.
Do I need to be on Ubuntu or some other OS to succeed with this?

I searched in the Discussions here, as well as elsewhere, but haven't seen anything yet. Any additional suggestions are welcome. Sorry if this is obvious.

Many thanks in advance.

**Comment by AlexPl292** [comment]
Hi there, I develop IdeaVim on MacOs.
There may be some linting issues, you can just ignore them. We perform a cleanup once a week.
The issue with `NoClassDefFoundError` seems to be some compilation issue. Maybe `./gradlew clean build` may help.
The issue with `File name too long` I haven't seen before :\ I suppose the previous command may also help.

Also, there are currently some issues with the IJ EAP, so it makes sense to change line 104 of `VimTestCase` to `    val fixtureBuilder = factory.createLightFixtureBuilder(projectDescriptor, "")
`.
Sorry for such inconvenience, we usually have a pretty simple development startup.

**Reply by jlangr** [reply]
Thanks Alex,

I realized I had JDK 11 at the command line, sorry. I switched to 17; with `./gradlew clean build` I get:

```
457 tests completed, 260 failed
```

That matches the 260 failures from running in IDEA.

The first test failure:
```
org.jetbrains.plugins.ideavim.RegisterActionsTest > initializationError FAILED
    java.lang.ExceptionInInitializerError at NativeConstructorAccessorImpl.java:-2
        Caused by: java.lang.reflect.InaccessibleObjectException at AccessibleObject.java:354
```

The remainder appear to be:
```
org.jetbrains.plugins.ideavim.action.AutoIndentTest > initializationError FAILED
    java.lang.NoClassDefFoundError at NativeConstructorAccessorImpl.java:-2
```

(I also tried with JDK 16. Sometimes the compile fails, by the way.)

I tried clearing my gradle cache as recommended by one StackOverflow post.

**Reply by citizenmatt** [reply]
I've got the same problem. I've changed the following line in `gradle.properties` to drop down to an earlier version of IDEA:

```
ideaVersion=2021.3
```

I can run tests again on Mac, now

**Reply by citizenmatt** [reply]
Actually, this might be due to the "file name too long" error. The `UsefulTestCase` base class will reset the global temp directory to a per-test folder during `setUp` and cleans it up again in `tearDown`. I've just had the problem where I was overriding `tearDown` and forgot to call `super.tearDown()`. This pushes another global temp directory for each test in the suite, and before you know it, there's a huge tree of folders in `$TMPDIR`. This also seems to cause unexpected side effects, with lots more exceptions. I guess if the temp dir isn't available (or is inaccessible because the entire path is way too long), then it destabilises the system.

Make sure you don't have any tests that override `tearDown` without calling `super`.

**Reply by jlangr** [reply]
>  I've changed the following line in `gradle.properties` to drop down to an earlier version of IDEA:
> 
> ```
> ideaVersion=2021.3
> ```

Thanks, I'd crossed my fingers but it didn't change the outcome. I tried also clearing my $TMPDIR among a few other things.

I don't see the `file name too long` error at this point, only the `java.lang.reflect.InaccessibleObjectException at AccessibleObject...` error and all the noclassdeffound exceptions that follow.

**Reply by jlangr** [reply]
Well I spoke too soon.

I did a gradle clean, then decided to muck in the IDE (all the prior attempts were at the command-line). Tests are all running now (they weren't earlier, so I'm presuming the ideaVersion made the difference).

Many thanks... I'll leave enough alone for the time being.
